invoice_modify_header/invoice.py

```diff
@@ -0,0 +1,125 @@
+                    ('//group[@id="buttons"]', 'states', {'invisible': True}),
+                    ])
+        return attributes
+
+    @classmethod
+    @ModelView.button_action('account_es.wizard_modify_header')
+    def modify_header(cls, invoices):
+        pass
+
+
+class ModifyHeaderStateView(StateView):
+    def get_view(self, wizard, state_name):
+        with Transaction().set_context(modify_header=True):
+            return super(ModifyHeaderStateView, self).get_view(
+                wizard, state_name)
+
+    def get_defaults(self, wizard, state_name, fields):
+        return {}
+
+
+class ModifyHeader(Wizard):
+    "Modify Header"
+    __name__ = 'invoice.modify_header'
+    start = ModifyHeaderStateView('account.invoice',
+        'account_es.modify_header_form', [
+            Button("Cancel", 'end', 'tryton-cancel'),
+            Button("Modify", 'modify', 'tryton-ok', default=True),
+            ])
+    modify = StateTransition()
+
+    def get_invoice(self):
+        if self.record.state != 'draft':
+            # TODO afegir message
+            raise AccessError(
+                gettext('account_es.msg_invoice_modify_header_draft',
+                    invoice=self.record.rec_name))
+        return self.record
+
+    def value_start(self, fields):
+        invoice = self.get_invoice()
+        values = {}
+        for fieldname in fields:
+            value = getattr(invoice, fieldname)
+            if isinstance(value, Model):
+                if getattr(invoice.__class__, fieldname)._type == 'reference':
+                    value = str(value)
+                else:
+                    value = value.id
+            elif isinstance(value, (list, tuple)):
+                value = [r.id for r in value]
+            values[fieldname] = value
+
+        # Mimic an empty invoice in draft state to get the fields' states right
+        values['lines'] = []
+        return values
+
+    def transition_modify(self):
+        pool = Pool()
+        Invoice = pool.get('account.invoice')
+        Line = pool.get('account.invoice.line')
+
+        invoice = self.get_invoice()
+        values = self.start._save_values
+
+        party_id = values.get('party')
+        # before save invoice, sure that the party line is the same
+        line_ids = []
+        for line in invoice.lines:
+            if line.party and line.party.id != party_id:
+                line_ids.append(line.id)
+        if line_ids:
+            table = Line.__table__()
+            cursor = Transaction().connection.cursor()
+            cursor.execute(*table.update(
+                    columns=[table.party],
+                    values=[party_id],
+                    where=table.id.in_(line_ids)))
+            invoice = Invoice(invoice.id)
+
+        self.model.write([invoice], values)
+        self.model.log([invoice], 'write', ','.join(sorted(values.keys())))
+
+        # Call on_change after the save to ensure parent invoice
```

Comment:
"before", not "after". Or the 'save' must be set the line above this.

Reply:
copy-paste from https://github.com/tryton/tryton/blob/55acbadcf240127a6bf1aecf93a86f6e097a9894/modules/sale/sale.py#L243

At line 116, save the invoice